Refactor julia initialization by finding and loading the sysimage earlier

gbaraldi opened the pull request:
This changes the order of initialization slightly. The goal is to be able to initialize executables and share libraries without having to explicitly call init functions.

This still doesn't address the thread safety

gbaraldi committed:
Refactor julia initialization by finding and loading the sysimage earlier

JeffBezanson commented on src/init.c:
```diff
@@ -721,8 +723,85 @@ static void restore_fp_env(void)
+        jl_gc_notify_image_load(sysimage.data, sysimage.size);
+
+    if (jl_options.cpu_target == NULL)
+        jl_options.cpu_target = "native";
```
Should this be the default value in jloptions.c?

topolarity commented:
Is this enough to fix `--trim` executables (c.f. https://github.com/JuliaLang/julia/pull/58197)?

topolarity committed:
Initializer timing sub-system earlier

topolarity commented on src/threading.c:
```diff
@@ -461,9 +461,7 @@ JL_DLLEXPORT jl_gcframe_t **jl_autoinit_and_adopt_thread(void)
             exit(1);
         }
-        const char *image_path = jl_pathname_for_handle(handle);
-        jl_enter_threaded_region(); // This should maybe be behind a lock, but it's harmless if done twice
```
Was this deletion intentional?

gbaraldi replied:
Yes. It got moved into the jl_init_with_handle call (the threaded region maybe should still be here :\ but I dunno)

topolarity replied:
Yeah it was specifically the thread region I was curious about - I'm OK with deleting it for now, we can re-visit soon with thread-safety

topolarity committed:
Fix `julia_bindir` calculation when embedding

topolarity committed:
Replace `jl_load_image_and_init` with inlined logic

This makes it clearer that sysimage resolution and `jl_options`
setup are the responsibility of `jl_api.c`, not the initialization
system.

topolarity committed:
Fix trimming test

topolarity commented:
I cannot abide CI any longer. I'm going to merge this as-is (it has passed 80% of CI several times now)

Pkg server issues have been worked around, but 5-10+ manual Windows restarts per run, random failures, and `cmdlineargs`-related failures (see https://buildkite.com/julialang/julia-master/builds/47119#01968dd3-efa0-414c-8aae-c2592c8a46cf/902-1488 and https://github.com/JuliaLang/julia/pull/58172#issuecomment-2846047822) make it almost impossible to get a clean run through.

topolarity commented:
Thanks @gbaraldi !
